src/loader/module_wrap.cc

```diff
@@ -0,0 +1,492 @@
+    return URL("");
+  }
+  URL resolve_module(std::string specifier, URL* base) {
+    URL parent = *base;
+    URL dir("");
+    do {
+      dir = parent;
+      auto check = Resolve("./node_modules/" + specifier, &dir, true);
+      if (!(check.flags() & URL_FLAGS_FAILED)) {
+        const auto limit = specifier.find('/');
+        std::string chroot = dir.path() + "node_modules/" + specifier.substr(
+          0,
+          limit == std::string::npos ? specifier.length() : limit + 1
+        );
+        if (check.path().substr(0, chroot.length()) != chroot) {
+          // fprintf(stderr, "tried to escape root\n");
+          return URL("");
+        }
+        return check;
+      }
+      else {
+        // TODO, PREVENT FALLTHROUGH
+      }
+      parent = URL("..", &dir);
+    } while (parent.path() != dir.path());
+    return URL("");
+  }
+  URL Resolve(std::string specifier, URL* base, bool read_pkg_json) {
+    auto pure_url = URL(specifier.c_str());
+    // printf("resolving, %s against %s\n", specifier.c_str(), base->path().c_str());
+    if (!(pure_url.flags() & URL_FLAGS_FAILED)) {
+      return pure_url;
+    }
+    if (specifier.length() == 0) {
+      return URL("");
+    }
+    if (is_relative_or_absolute_path(specifier)) {
+      auto resolved = URL(specifier, base);
+      auto file = resolve_extensions(resolved);
+      if (!(file.flags() & URL_FLAGS_FAILED)) return file;
+      if (specifier.back() != '/') {
+        resolved = URL(specifier + "/", base);
+      }
+      if (read_pkg_json) {
+        auto main = resolve_main(resolved);
+        if (!(main.flags() & URL_FLAGS_FAILED)) return main;
+      }
+      return resolve_index(resolved);
+    }
+    else {
+      return resolve_module(specifier, base);
+    }
+    return URL("");
+  }
+
+void ModuleWrap::Resolve(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  if (args.IsConstructCall()) {
+    env->ThrowError("resolve() must not be called as a constructor");
+    return;
+  }
+  if (args.Length() != 2) {
+    env->ThrowError("resolve must have exactly 2 arguments (string, string)");
+    return;
+  }
+
+  if (!args[0]->IsString()) {
+    env->ThrowError("first argument is not a string");
+    return;
+  }
+  auto specifier_arg = args[0]->ToString();
+  String::Utf8Value specifier_utf(specifier_arg);
+
+  if (!args[1]->IsString()) {
+    env->ThrowError("second argument is not a string");
+    return;
+  }
+  auto url_arg = args[1]->ToString();
+  String::Utf8Value url_utf(url_arg);
+  URL url(*url_utf);
+  // printf("resolving, args.Length() %d %s\n", args.Length(), *url_utf);
+  if (url.flags() & URL_FLAGS_FAILED) {
+    env->ThrowError("second argument is not a URL string");
+    return;
+  }
+
+  URL result = node::loader::Resolve(*specifier_utf, &url, true);
+  if (result.flags() & URL_FLAGS_FAILED) {
+    env->ThrowError("module not found");
+    return;
+  }
+
+  const Local<Value> ret(result.ToObject(
```

Comment:
@TimothyGu I agree with your comments here, but it might be easiest to just do what @targos did and push changes, esp. for suggestions like these, yourself